### PR TITLE
feat: adding abstract, readonly and additional public/protected/private groups

### PR DESCRIPTION
**Pull request by hugop95**
Added the additional groups:

`'public-abstract-property'`
`'protected-abstract-property`'
`'abstract-property'`
`'public-abstract-method'`
`'protected-abstract-method'`
`'abstract-method'`
`'static-public-property'`
`'static-private-property'`
`'static-public-method'`
`'public-constructor'`
`'protected-constructor'`
`'private-constructor'`
`'public-property'`
`'public-method'`

### CODE REVIEW

**Comment by hugop95** [comment]
@azat-io As mentioned in https://github.com/azat-io/eslint-plugin-perfectionist/issues/190,
I am adding `abstract` groups as well as some other missing groups that I think would be useful. We could also add `readonly` as per https://github.com/azat-io/eslint-plugin-perfectionist/issues/193.

Do you think the number of official groups would be too much?

File: .gitignore [hunk 1]
```diff
@@ -17,3 +18,6 @@ coverage/
 
 # Misc
 .DS_Store
+
+# Jetbrains IDE
+.idea
```
Comment (azat-io) [inline .gitignore:23]: Could you add `.idea` to `# Editor` block?

File: .gitignore [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # Node.js
 node_modules/
 .nvmrc
+package-lock.json
```
Comment (azat-io) [inline .gitignore:4]: We use pnpm and we don't need to ignore the lock file for npm

**Review by azat-io — COMMENTED** [review]
Thank you for your pull request! It looks good.

Could you please also fix the following issues:

1. Add new groups to docs (`docs/content/rules/sort-classes.mdx`)
2. Fix ESLint issues
3. Fix Unit tests
4. Add tests for new groups. We want to have 100% code coverage with tests. To run tests, use the `pnpm run test:unit` command.

**Comment by hugop95** [comment]
@azat-io The number of conditional branches is growing exponentially as we allow new modifiers to be added (in this case, `abstract` and `readonly`. We increase that number even more if we choose to allow for other keywords like `override` or `async` in the future.

What do you think about instead dynamically building a list of `modifiers`, and then constructing the list of groups associated?

Example:

`private abstract get a() {...`
Should generate the modifiers `['private', 'abstract']`

Associated to the selectors `['get-method', 'method']`

From which we can generate generate:
`['private-abstract-get-method', 'private-abstract-method', 'abstract-get-method', 'abstract-method', 'private-get-method', 'private-method', 'get-method', 'method']

**Comment by azat-io** [comment]
Yes, the growing number of groups looks like a problem.

What should the API for modifiers look like? I don't really understand the idea yet.

**Comment by hugop95** [comment]
I don't think we have to change the API yet. However, in the code, instead of doing 

```
            let memberGroups = []

            ...

            if (isPrivateProperty) {
                    if (isReadonly) {
                      if (decorated) {
                        memberGroups.push('private-readonly-decorated-property')
                      }
                      memberGroups.push('private-readonly-property')
                    }
                }
            }
           
            // Other complex branches

            for (let memberGroup of memberGroups) {
              defineGroup(memberGroup)
            }
```
We could do:
```
    let modifiers = [];
    let selectors = [];
    if (isStaticMethod) {
      improvedMemberGroups.push('static');
    }
    if (isPrivateMethod) {
      modifiers.push('private');
    } else if (isProtectedMethod) {
      modifiers.push('protected');
    } else {
      modifiers.push('public');
    }
    if (isAbstractMethod) {
      modifiers.push('abstract');
    }
    if (decorated) {
      modifiers.push('decorated');
    }
    if (isGetMethod) {
      selectors.push('get-method');
    } else if (isSetMethod) {
      selectors.push('set-method');
    } 
    selectors.push('method');
    }

   // Generate memberGroups from modifiers + selectors
   let memberGroups = generateMemberGroups(modifiers, selectors)
    for (let memberGroup of memberGroups) {
          defineGroup(memberGroup)
     }
```

**Comment by hugop95** [comment]
My idea of the `generateMemberGroups` function is to iterate over all modifiers union and add selectors at the end, like the following example:

element: `private abstract get a();`
modifiers: `['private', 'abstract']`
selectors: ['get-method', 'method']`

=> `['private-abstract-get-method', 'private-abstract-method', 'abstract-get-method', 'abstract-method', 'private-get-method', 'private-method', 'get-method', 'method']`

If we want to be lenient and allow users to enter `abstract-private-get-method` instead of `private-abstract-get-method`, we can make it so the `generateMemberGroups(modifiers, selectors)` generate all the possible combinations possible regardless of order.

**Comment by hugop95** [comment]
I consider https://github.com/azat-io/eslint-plugin-perfectionist/pull/196 to be a superior implementation compared to this.

**Comment by azat-io** [comment]
Closed in favor of #196.

**Pull request closed without merging**